### DISCUSSION TITLE
What is the best practice to handle multiple actions in the same page?

### DISCUSSION THREAD

**Discussion by VigneshVaidyanathan** [remix-run/remix#3138] [Q&A] [answered]
I am currently having a page that has these many actions that the user can do
1. Upload a file
2. Add a new todo item
3. Delete an existing item
4. Mark an item as done
5. Mark an item as undone

The process that I am currently following is that the page has a top action function that handles all these actions. (all the actions are sent by using fetchers). So inside the action I try and get the form values where every interaction has a separate `_action` property that is set to the type of action i.e. `upload-image`, `delete-item`, `mark-done`, `mark-undone` etc

The action function is quite big now to handle all these cases. There are two ways that I can reduce this and split it into different functions.
1. Have a simple switch and split the actual logic to handle the action in a separate file that does the job. So the action just has switch and calls the necessary function.
2. Create different actions for each interaction like `api/todo/upload-image`, `api/todo/delete` etc so that can be directly called by setting the right path but I need to restrict the user from typing the URL on the browser that matches this path. 

Is there a preferred way to handle multiple actions? Am I doing something very wrong or missing something that is very basic?

**Comment by kiliman** [comment, marked as answer]
Option 1 would be the simplest and acceptable solution. This is standard code decomposition. 

The beauty of having a single endpoint for your actions is that you can immediately tell what the route does. If you have many API endpoints, then you're kind of back to the old way of building React apps.

But how you structure your code is a personal preference. I'd do the simplest thing possible then refactor if it starts to become an issue.

**Reply by VigneshVaidyanathan** [reply]
Sure, I was thinking the same but encountered a blog where they did the option 2. So just had to confirm if that has any advantage in Remix or some convention.

Thanks again for the answer.

**Reply by kiliman** [reply]
Option 2 is probably more familiar to people, so I can see why they may have chosen that by default.

**Reply by mikkpokk** [reply]
@kiliman Actually one and only reason why I have chosen option 2 in some cases, is because every `action` allows to post only 1 request at the time. If user happens to click another button, which will trigger the same action but different outcome (ie. upload image & resort images) then the "upload image" task will be cancelled.

I haven't found that Remix has any option to avoid such a behaviour without creating separate action endpoints.

**Reply by sergiodxa** [reply]
> every action allows to post only 1 request at the time

This is not entirely correct.

Form and useSubmit will abort each other yes, so if the user submit's a Form and then another one it will abort the first one, note that "abort" here means the browser will not receive the response and Remix will ignore it, the server already received the request and unless you use `request.signal` to stop processing it, it will not do it for you. This behavior also includes useSubmit, so if call the submit function returned by the hook two times the first will be aborted, and Form and useSubmit will do the same between each other, this is because Form uses the useSubmit hook internally.

But if you use the useFetcher hook, each fetcher instance will work independently of each other, this means that if you use `useFetcher().submit` and/or `useFetcher().Form` with the same hook instance the second request will abort the first one, but if you call useFetcher two times, and each one does a separate request they will not abort each other.

So if you have something like a list of posts the user can like, each post can have a useFetcher instance and be liked/disliked independently without aborting the other likes/dislikes.

Note also that all of this happens client-side, so the `action` function itself has no limit to handle only one request at the time, this is controlled by Form, useSubmit and useFetcher.

**Comment by ckuijjer** [comment]
In [Add Delete Functionality to Posts Page in Remix](https://egghead.io/lessons/remix-add-delete-functionality-to-posts-page-in-remix) Kent Dodds adds a `name="intent"` and e.g. `value="delete"` to each of the `<button />`'s in the form and in the action function use `formData.get('intent')` to decide on the actual logic. This makes me feel that option 1 would be the common way to go

**Reply by luispuig** [reply]
I am doing some experiments on this issue. remix-tracks is on of them: https://github.com/luispuig/remix-tracks
Hopefully we can a solution that works for many use-cases
